File: adapters/media/thumbnail/style.go
package thumbnail

import (
	"image"
	"image/color"
)

// Every tunable of the thumbnail's look, in one file.
//
// Nothing here is logic: it is the design expressed as numbers, so it can be
// changed by editing this file rather than by reading the drawing code. Each
// entry says what raising it does.
//
// They are constants rather than settings rows on purpose. The backend that
// exists to be re-styled without a rebuild is the browser one slide to come;
// a settings table full of pixel values would be a second place to change a
// layout and a first place to get it wrong.

// ------------------------------------------------------------------ frame ---

const (
	// YouTube's thumbnail size.
	frameWidth  = 1280
	frameHeight = 720

	// backgroundBrightness is how much of the backdrop photograph survives, out
	// of 255. Lower is darker: white type over an undimmed photograph is
	// unreadable, and the reference thumbnails are nearly black behind it.
	backgroundBrightness = 220
)

// ------------------------------------------------------------------- grid ---

// The grid is sized from the frame width first and the headline takes what is
// left over, which is what keeps the tiles running edge to edge.
const (
	// gridSideMargin is the gutter on each side of the grid. Lower it for a
	// wider grid; zero puts the tiles hard against the frame.
	gridSideMargin = 24
	// gridBottomMargin is the room under the last row of captions.
	gridBottomMargin = 14
	// headlineToGridGap separates the headline from the first row of tiles.
	headlineToGridGap = 18
	// tileSpacing is the gap between tiles, across and down.
	tileSpacing = 28
)

// ------------------------------------------------------------------- tile ---

const (
	// tileBorderWidth is the frame drawn around each tile. Zero for none.
	tileBorderWidth = 3
	// tileCornerRadius rounds the tile's corners, the border and the icon inside
	// it together. Zero is square. It is clamped to half the tile, so a large
	// value simply rounds a tile into a circle rather than misdrawing it.
	tileCornerRadius = 10
	// tileIconPadding is the inset from the border to the icon. Raise it to give
	// the artwork more air.
	tileIconPadding = 8
)

// ---------------------------------------------------------------- caption ---

// The caption type size is chosen once for the whole grid — the largest between
// the bounds at which every caption fits its tile. Set the two equal to pin the
// size and let long captions be cut instead.
const (
	// tileToCaptionGap is the space between a tile and the caption under it.
	tileToCaptionGap = 8
	captionFontMax   = 26
	captionFontMin   = 12
	// captionFontStep is how finely the fitting walks between those two.
	captionFontStep = 1
	// captionRowHeight is the height reserved under every tile whatever size the
	// captions settle at. Reserving the same band for all of them is what keeps
	// the rows aligned, so it wants to be a little over captionFontMax.
	captionRowHeight = captionFontMax + 2
)

// --------------------------------------------------------------- headline ---

// The headline is fitted into whatever height the grid left above it, at the
// largest size between the bounds that fits on one line. It wraps only when
// even the floor will not hold it.
//
// A fuller grid leaves the headline less room: the default twelve tiles in two
// rows of six sit near the middle of this range.
const (
	headlineSideMargin = 40
	headlineTopMargin  = 22
	headlineFontMax    = 120
	headlineFontMin    = 44
	// headlineFontStep is how finely the fitting walks between those two.
	headlineFontStep = 4
	// headlineLineGap is the space between wrapped lines.
	headlineLineGap = 6
	// headlineMaxLines is the most lines a headline may take before it is drawn
	// at the floor and allowed to overflow.
	headlineMaxLines = 2
	// headlineTracking is the letter-spacing, as a divisor of the type size:
	// spacing = size / headlineTracking. Lower is looser.
	headlineTracking = 22
)

// ------------------------------------------------------------------- icon ---

// The icons arrive as artwork on a solid dark field — that is what an image
// model returns when asked for line art, and what the samples are. Drawn as-is
// they would stamp that field over the tile, so the darkness becomes
// transparency instead. The ramp between the two thresholds is what stops the
// artwork's anti-aliased edges from turning into a hard jagged cut.
const (
	// iconTransparentBelow: luminance at or under this disappears entirely.
	// Raise it to eat a background that is not quite black.
	iconTransparentBelow = 48
	// iconOpaqueAbove: luminance at or over this is kept as-is. Lower it to keep
	// more of the artwork's darker shading.
	iconOpaqueAbove = 105
)

// ---------------------------------------------------------------- palette ---

// Colours are vars only because Go has no constant structs; treat them the same
// way. Alpha is honoured everywhere: a tileFillColor of A: 0 lets the backdrop
// show through the tiles entirely, and A: 255 makes them solid plates.
var (
	// headlineColor is the hook across the top; captionColor the labels under
	// each tile.
	headlineColor = image.NewUniform(color.RGBA{R: 246, G: 246, B: 244, A: 255})
	captionColor  = image.NewUniform(color.RGBA{R: 226, G: 226, B: 222, A: 255})

	// tileFillColor is the plate behind each icon. tileBorderColor is a hair off
	// pure white: at 255 it out-shouts the artwork it frames.
	tileFillColor   = color.RGBA{R: 6, G: 6, B: 8, A: 235}
	tileBorderColor = color.RGBA{R: 228, G: 228, B: 224, A: 255}
)

// -------------------------------------------------------------- resources ---

const (
	// backgroundFileName is read from the resources directory; defaultFontFile
	// from its fonts subdirectory. The font is also a settings row, so this is
	// only the fallback.
	backgroundFileName = "background.jpg"
	defaultFontFile    = "CabinSketch-Bold.ttf"
	// defaultGridRows applies when the settings row is unreadable.
	defaultGridRows = 2
)

File: web/src/v2/components/thumbnail/style.ts
/**
 * Every tunable of the thumbnail's look, mirrored from the Go renderer, and
 * turned into a value the operator can change.
 *
 * `adapters/provider/thumbnail/style.go` is the original: a block of constants
 * with a note on each saying what raising it does. Its own header explains why
 * they are constants there — "the backend that exists to be restyled without a
 * rebuild is the HTML one still to come" — and this is that backend. So the
 * numbers are the same numbers, transcribed entry for entry, but they are
 * fields on a `Style` rather than compile-time constants.
 *
 * `defaultStyle` is exactly what the Go renderer uses. That matters more than
 * it looks: a thumbnail built here without touching a control composes the
 * image the pipeline would have composed, so the builder is a place to *check*
 * the render before it publishes and not only a place to depart from it. The
 * moment a control moves, the two diverge — which is legitimate, because an
 * override is by definition an image the renderer would not have made, but it
 * is a thing the operator should be told rather than left to discover.
 *
 * When style.go changes, this changes. There is no clever way to share them:
 * one is compiled into a binary and the other is shipped to a browser, and
 * generating this file would put a build step between a designer and a number.
 */

/* ------------------------------------------------------------------ frame */

/**
 * YouTube's thumbnail size, from `entity.ThumbnailWidth/Height`.
 *
 * Not on `Style`, because it is not a look: an image outside these is refused
 * at upload whatever drew it. That is why they live in the entity on the Go
 * side rather than among the renderer's style constants.
 */
export const frameWidth = 1280
export const frameHeight = 720

/**
 * How finely the fitting loops walk between their bounds.
 *
 * Also not on `Style`. These change how long the search takes and which of two
 * adjacent sizes it lands on, not what the design is, and a control for them
 * would be a control for the algorithm rather than for the picture.
 */
export const captionFontStep = 1
export const headlineFontStep = 4

/* ------------------------------------------------------------------ style */

export interface Style {
  /** A filename in the resources fonts directory. */
  font: string
  /** How many rows the grid is laid out in. */
  rows: number

  /* grid — sized from the frame width first; the headline takes what is left */

  /** The gutter each side of the grid. Zero puts tiles against the frame. */
  gridSideMargin: number
  /** The room under the last row of captions. */
  gridBottomMargin: number
  /** Separates the headline from the first row of tiles. */
  headlineToGridGap: number
  /** The gap between tiles, across and down. */
  tileSpacing: number

  /* tile */

  /** The frame drawn around each tile. Zero for none. */
  tileBorderWidth: number
  /** Rounds the tile, its border and its icon together. Zero is square. */
  tileCornerRadius: number
  /** The inset from the border to the icon. Raise it for more air. */
  tileIconPadding: number
  /** The plate behind each icon. */
  tileFillColor: string
  /** How opaque that plate is, out of 255. At 0 the backdrop shows through. */
  tileFillAlpha: number
  /** A hair off pure white: at 255 it out-shouts the artwork it frames. */
  tileBorderColor: string
  /**
   * How opaque that border is, out of 255.
   *
   * The Go renderer has no equivalent field: the alpha is carried by
   * `tileBorderColor` itself, which is declared there as an RGBA and blended by
   * the painter at whatever alpha it holds. So the machinery for a translucent
   * border was always in place and only the number was fixed. 204 is that
   * number — four fifths — which is why leaving this alone still composes what
   * the renderer composes.
   */
  tileBorderAlpha: number

  /* caption — one size for the whole grid, the largest at which all fit */

  /** The space between a tile and the caption under it. */
  tileToCaptionGap: number
  captionFontMax: number
  captionFontMin: number
  captionColor: string

  /* headline — fitted into whatever height the grid left above it */

  headlineSideMargin: number
  headlineTopMargin: number
  headlineFontMax: number
  headlineFontMin: number
  /** The space between wrapped lines. */
  headlineLineGap: number
  /** The most lines before it is drawn at the floor and allowed to overflow. */
  headlineMaxLines: number
  /** Letter-spacing as a divisor of the type size: spacing = size / this. */
  headlineTracking: number
  headlineColor: string
  /**
   * The headline's function words, drawn in `headlineMinorColor` rather than
   * at full white so the words that carry the hook read first.
   *
   * A string rather than a parsed list, because it mirrors a settings row that
   * an operator types into -- `thumbnail.headline.minor_words` -- and the
   * separators it forgives, commas or whitespace, are forgiven here too.
   * Empty draws the whole headline in `headlineColor`.
   */
  headlineMinorWords: string
  /**
   * A dimmer neutral rather than a second hue, which is the whole design: the
   * gap the eye resolves at browse size is a luminance gap, and a hue-only
   * demotion collapses at 320px.
   *
   * Tied to `headlineColor` and not readable alone — dragging that swatch to a
   * darker hue without bringing this one down leaves the demoted words brighter
   * than the ones they are demoted against. See `headlineMinorColor` in
   * style.go for the numbers.
   */
  headlineMinorColor: string

  /* image */

  /** How much of the backdrop survives, out of 255. Lower is darker. */
  backgroundBrightness: number
  /** Luminance at or under this disappears entirely from an icon. */
  iconTransparentBelow: number
  /** Luminance at or over this is kept as-is. */
  iconOpaqueAbove: number
}

/** Exactly what the Go renderer uses. See the note at the top of the file. */
export const defaultStyle: Style = {
  font: 'CabinSketch-Bold.ttf',
  rows: 2,

  gridSideMargin: 24,
  gridBottomMargin: 14,
  headlineToGridGap: 18,
  tileSpacing: 28,

  tileBorderWidth: 3,
  tileCornerRadius: 10,
  tileIconPadding: 8,
  tileFillColor: '#060608',
  tileFillAlpha: 235,
  tileBorderColor: '#e4e4e0',
  tileBorderAlpha: 204,

  tileToCaptionGap: 8,
  captionFontMax: 26,
  captionFontMin: 12,
  captionColor: '#e2e2de',

  headlineSideMargin: 40,
  headlineTopMargin: 22,
  headlineFontMax: 120,
  headlineFontMin: 44,
  headlineLineGap: 6,
  headlineMaxLines: 2,
  headlineTracking: 22,
  headlineColor: '#ff2600',
  // The seeded value of `thumbnail.headline.minor_words`, from
  // `entity.DefaultHeadlineMinorWords`.
  headlineMinorWords:
    'a,an,and,are,as,at,be,but,by,for,from,in,is,of,on,or,so,than,that,the,this,to,was,were,with',
  headlineMinorColor: '#585854',

  backgroundBrightness: 220,
  iconTransparentBelow: 48,
  iconOpaqueAbove: 105,
}

/**
 * The height reserved under every tile, whatever size the captions settle at.
 *
 * Derived rather than stored, exactly as in style.go where it is
 * `captionFontMax + 2`. Reserving the same band for all of them is what keeps
 * the rows aligned, so it wants to be a little over the ceiling.
 */
export function captionRowHeight(style: Style): number {
  return style.captionFontMax + 2
}

/** Whether anything has been moved off what the renderer would have used. */
export function isDefault(style: Style): boolean {
  return (Object.keys(defaultStyle) as (keyof Style)[]).every(
    (key) => style[key] === defaultStyle[key],
  )
}

/* -------------------------------------------------------------- resources */

export const backgroundURL = '/resources/background.jpg'

/** Where a typeface is served from, by the name the settings row carries. */
export function fontURL(file: string): string {
  return `/resources/fonts/${file}`
}

/* ---------------------------------------------------------------- colours */

/** `#rrggbb` and an alpha out of 255, as the channels the painter blends. */
export function rgba(hex: string, alpha: number): { r: number; g: number; b: number; a: number } {
  const value = hex.replace('#', '')
  return {
    r: parseInt(value.slice(0, 2), 16) || 0,
    g: parseInt(value.slice(2, 4), 16) || 0,
    b: parseInt(value.slice(4, 6), 16) || 0,
    a: alpha,
  }
}
